bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java

```diff
@@ -209,6 +217,42 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
 		}
 	}
 
+	@Inject
+	@Optional
+	void subscribePerspectiveSwitched(
+			@UIEventTopic(UIEvents.UILifeCycle.PERSPECTIVE_SWITCHED) org.osgi.service.event.Event event) {
+		Object element = event.getProperty(EventTags.ELEMENT);
+		if (element instanceof MPerspective) {
+			setOnboarding((MPerspective) element);
+		}
+	}
+
+	private void setOnboarding(MPerspective perspective) {
+		if (onBoarding == null) {
+			return;
+		}
+
+		onBoarding.getDisplay().asyncExec(() -> {
```

Comment:
Why is this code runs in asyncExec? Please note. `asyncExec()` might run at any time in the future, even if the perspective already closed, but there are no checks at all inside the body if the code still should run.

Reply:
I don't remember. Was a back and forth during development.
I removed it. Let's see whether the tests are working now.